# Task#9. Titanic

# Текст задачи

## Задача 9. Титаник

* **Дедлайн**: 28.11.2024, 23:59
* Полный балл: 5

## Задача

Во время семинара на датасете [Титаник](https://www.kaggle.com/competitions/titanic/data) были получены следующие значение ошибок
- Для `DecisionTreeClassifier`
  - `precision` для выживших --- 0.79
  - `recall` для выживших --- 0.72
  - `f1_score` для выживших --- 0.75
- Для `RandomForestClassifier`
  - `accuracy` --- 0.8
- Для `LogisticRegression`
  - `precision` для погибших --- 0.81
  - `recall` для погибших --- 0.88
- Для `LinearRegression`
  - `RootMeanSquaredError` для возраста пассажиров --- 10.7


- [x] Используя любой из классификаторов получить значения `precision`, `recall` и `accuracy` выше, чем у моделей с занятия.
Если не удалось получить результат лучше, провести подробный анализ причины неудачи.
  - [x] Считать данные, выполнить первичный анализ данных, при необходимости произвести чистку данных (**Data Cleaning**).
  - [x] Выполнить разведочный анализ (**EDA**), использовать визуализацию, сделать выводы, которые могут быть полезны при дальнейшем решении задачи классификации.
  - [x] При необходимости выполнить полезные преобразования данных (например, трансформировать категориальные признаки в количественные), убрать ненужные признаки, создать новые (**Feature Engineering**).
- [x] Используя любую из регрессионных задач получить значение `RootMeanSquaredError` ниже, чем у модели с занятия.
Если не удалось получить результат лучше, провести подробный анализ причины неудачи.
  - [x] Считать данные, выполнить первичный анализ данных, при необходимости произвести чистку данных (**Data Cleaning**).
  - [x] Выполнить разведочный анализ (**EDA**), использовать визуализацию, сделать выводы, которые могут быть полезны при дальнейшем решении задачи классификации.
  - [x] При необходимости выполнить полезные преобразования данных (например, трансформировать категориальные признаки в количественные), убрать ненужные признаки, создать новые (**Feature Engineering**).
- [x] Сделать выводы о полученных результатах.


## $\blacksquare$ Считать данные, выполнить первичный анализ данных, при необходимости произвести чистку данных (**Data Cleaning**).

### Получение данных

Считаем данные, предварительно загруженные на Диск с Kaggle.

In [1]:
# import os
# from google.colab import drive

In [2]:
# # Connect Google Drive to the Collab
# gdrive_dir = '/content/drive/'

# drive.mount(gdrive_dir, force_remount = True)

Mounted at /content/drive/


In [3]:
import pandas as pd

In [4]:
!ls # Сheck where I am now

drive  sample_data


In [5]:
# Read data
train_df = pd.read_csv('/content/drive/MyDrive/spbu-python-course/task#9/titanic_dataset/train.csv')
test_df = pd.read_csv('/content/drive/MyDrive/spbu-python-course/task#9/titanic_dataset/test.csv')

### Первичный анализ данных

Посмотрим первые записи из каждого датафрейма, чтобы узнать его структуру.

In [6]:
# Viewing first rows of each dataframe
display(train_df.head())
display(test_df.head())

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

Узнаем общее число ненулевых значений. Будем использовать `info()`, `isnull().sum()`.

In [7]:
display(train_df.info())
display(test_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


None

In [8]:
display(train_df.isnull().sum())
display(test_df.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Как видно, некторые значения в столбцах `Age`, `Cabin`, `Embarked` для `train_df` отсутствуют.


Также, отсутствуют `Age`, `Fare`, `Cabin`, для `test_df`.

Узнаем общее описание числовых полей датафреймов с помощью `describe`

In [9]:
display(train_df.describe())
display(test_df.describe())

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

Из вывода выше, видно, что на тренировочном фрейме в среднем число выживших(наше ключевое поле для анализа) из 891 пассажира на Титанике составило 38%. В дальнешем, мы должны определить, от чего зависит это значение для пассажиров.

### Чистка данных

Заполним пропущенные значения:

`Age` - медианой (а не средним арифметическим например, т.к. оно очень чувствительно к выбросам).

`Embarked` - модой.

`Fare` - медианой.


`Cabin` - заполним '1' если есть, '0' - если отсутствует

In [10]:
train_df['Age'] = train_df['Age'].fillna(train_df['Age'].median())
test_df['Age'] = test_df['Age'].fillna(test_df['Age'].median())
train_df['Embarked'] = train_df['Embarked'].fillna(train_df['Embarked'].mode()[0])
test_df['Fare'] = test_df['Fare'].fillna(test_df['Fare'].median())
train_df['Cabin'] = train_df['Cabin'].notnull().astype(int)
test_df['Cabin'] = test_df['Cabin'].notnull().astype(int)

In [11]:
# Вывод полученных датасетов
print("Тренировочный набор данных после обработки:")
display(train_df.head())

print("Тестовый набор данных после обработки:")
display(test_df.head())

Тренировочный набор данных после обработки:


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare  Cabin Embarked  
0      0         A/5 21171   7.2500      0        S  
1      0          PC 17599  71.2833      1        C  
2      0  STON/O2. 3101282   7.9250      0        S  
3      0            113803  53.1000      1        S  
4      0            373450   8.0500      0        S

Тестовый набор данных после обработки:


PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare  Cabin Embarked  
0  34.5      0      0   330911   7.8292      0        Q  
1  47.0      1      0   363272   7.0000      0        S  
2  62.0      0      0   240276   9.6875      0        Q  
3  27.0      0      0   315154   8.6625      0        S  
4  22.0      1      1  3101298  12.2875      0        S

Проверим данные после обработки:

In [12]:
display(train_df.isnull().sum())
display(test_df.isnull().sum())

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

Теперь пропущенные значения отсутствуют

## $\blacksquare$При необходимости выполнить полезные преобразования данных (например, трансформировать категориальные признаки в количественные), убрать ненужные признаки, создать новые (**Feature Engineering**).

### Убрать лишние признаки

Удалим лишние(сложные для интерпретации) поля такие как `Name`, `Ticket`

In [13]:
train_df.drop(['Name', 'Ticket'], axis=1, inplace=True)
test_df.drop(['Name', 'Ticket'], axis=1, inplace=True)

### Трансформируем категориальные признаки в колличественные

Заменим категориальные поля `Sex`, `Embarked`, на числовые

In [14]:
train_df['Sex'] = train_df['Sex'].map({'male': 0, 'female': 1})
test_df['Sex'] = test_df['Sex'].map({'male': 0, 'female': 1})

train_df['Embarked'] = train_df['Embarked'].map({'C': 0, 'Q': 1, 'S': 2})
test_df['Embarked'] = test_df['Embarked'].map({'C': 0, 'Q': 1, 'S': 2})

In [15]:
# Вывод полученных датасетов
print("Тренировочный набор данных после обработки:")
display(train_df.head())

print("Тестовый набор данных после обработки:")
display(test_df.head())

Тренировочный набор данных после обработки:


PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Cabin  \
0            1         0       3    0  22.0      1      0   7.2500      0   
1            2         1       1    1  38.0      1      0  71.2833      1   
2            3         1       3    1  26.0      0      0   7.9250      0   
3            4         1       1    1  35.0      1      0  53.1000      1   
4            5         0       3    0  35.0      0      0   8.0500      0   

   Embarked  
0         2  
1         0  
2         2  
3         2  
4         2

Тестовый набор данных после обработки:


PassengerId  Pclass  Sex   Age  SibSp  Parch     Fare  Cabin  Embarked
0          892       3    0  34.5      0      0   7.8292      0         1
1          893       3    1  47.0      1      0   7.0000      0         2
2          894       2    0  62.0      0      0   9.6875      0         1
3          895       3    0  27.0      0      0   8.6625      0         2
4          896       3    1  22.0      1      1  12.2875      0         2

### Создание новых

Создадим новый признак `Family`, представляющий сумму `Parch` и `SibSp`. Так как по моему предположению, пассажир действительно мог путешествовать с семьей, которая так и задавалась, а эти данные разделили на два признака в датафрейме.
Обьединяя их, мы сократим число признаков, упростив общую структуру данных.

In [16]:
train_df['Family'] = train_df['Parch'] + train_df['SibSp']
test_df['Family'] = test_df['Parch'] + test_df['SibSp']
train_df.drop(['Parch', 'SibSp'], axis=1, inplace=True)
test_df.drop(['Parch', 'SibSp'], axis=1, inplace=True)

In [17]:
# Вывод полученных датасетов
print("Тренировочный набор данных после обработки:")
display(train_df.head())

print("Тестовый набор данных после обработки:")
display(test_df.head())

Тренировочный набор данных после обработки:


PassengerId  Survived  Pclass  Sex   Age     Fare  Cabin  Embarked  Family
0            1         0       3    0  22.0   7.2500      0         2       1
1            2         1       1    1  38.0  71.2833      1         0       1
2            3         1       3    1  26.0   7.9250      0         2       0
3            4         1       1    1  35.0  53.1000      1         2       1
4            5         0       3    0  35.0   8.0500      0         2       0

Тестовый набор данных после обработки:


PassengerId  Pclass  Sex   Age     Fare  Cabin  Embarked  Family
0          892       3    0  34.5   7.8292      0         1       0
1          893       3    1  47.0   7.0000      0         2       1
2          894       2    0  62.0   9.6875      0         1       0
3          895       3    0  27.0   8.6625      0         2       0
4          896       3    1  22.0  12.2875      0         2       2

## $\blacksquare$ Выполнить разведочный анализ (**EDA**), использовать визуализацию, сделать выводы, которые могут быть полезны при дальнейшем решении задачи классификации.

### EDA

Вновь посмотрим на основные характеристики тренировочного датафрейма

In [18]:
# Основные статистики
display(train_df.describe())
# Информация о типах данных и пропусках
display(train_df.info())
# Проверка на пропуски
display(train_df.isnull().sum())

PassengerId    Survived      Pclass         Sex         Age  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642    0.352413   29.361582   
std     257.353842    0.486592    0.836071    0.477990   13.019697   
min       1.000000    0.000000    1.000000    0.000000    0.420000   
25%     223.500000    0.000000    2.000000    0.000000   22.000000   
50%     446.000000    0.000000    3.000000    0.000000   28.000000   
75%     668.500000    1.000000    3.000000    1.000000   35.000000   
max     891.000000    1.000000    3.000000    1.000000   80.000000   

             Fare       Cabin    Embarked      Family  
count  891.000000  891.000000  891.000000  891.000000  
mean    32.204208    0.228956    1.536476    0.904602  
std     49.693429    0.420397    0.791503    1.613459  
min      0.000000    0.000000    0.000000    0.000000  
25%      7.910400    0.000000    1.000000    0.000000  
50%     14.454200    0.000000    2.000000    0.000000  
75%     31.000000    0.000000    2.000000    1.000000  
max    512.329200    1.000000    2.000000   10.000000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    int64  
 4   Age          891 non-null    float64
 5   Fare         891 non-null    float64
 6   Cabin        891 non-null    int64  
 7   Embarked     891 non-null    int64  
 8   Family       891 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 62.8 KB


None

PassengerId    0
Survived       0
Pclass         0
Sex            0
Age            0
Fare           0
Cabin          0
Embarked       0
Family         0
dtype: int64

На данный момент можем выдвинуть гипотезу о том, что женщины зачастую выживали чаще, и если у пассажира была каюта(билет дороже) то, и шанс на его выживаение тоже больше. Попробуем это проверить.

### Визуализация

**Анализ распределения целевой переменной (`Survived`)**

Рассмотрим распределение выживших и погибших:

In [19]:
import plotly.express as px
import numpy as np
import plotly.graph_objects as go

In [20]:
# Преобразуем результаты value_counts() в DataFrame с правильными колонками
survived_counts = train_df['Survived'].value_counts().reset_index()
survived_counts.columns = ['Survived', 'count']  # Переименовываем колонки

# Создаем интерактивный график
fig = px.bar(
    survived_counts,
    x='Survived',
    y='count',
    color='Survived',
    labels={'Survived': 'Выживаемость (0 = Нет, 1 = Да)', 'count': 'Количество'},
    title='Распределение выживших',
    color_discrete_sequence=px.colors.sequential.Viridis
)
fig.show()

**Влияние пола на выживаемость(`Sex`)**


Посмотрим, как пол пассажира влияет на выживаемость:

In [21]:
# Группируем данные по полу и считаем среднюю выживаемость
sex_survived = train_df.groupby('Sex')['Survived'].mean().reset_index()

# Создаем интерактивный график
fig = px.bar(
    sex_survived,
    x='Sex',
    y='Survived',
    color='Sex',
    labels={'Sex': 'Пол (0 = Мужчина, 1 = Женщина)', 'Survived': 'Доля выживших'},
    title='Выживаемость в зависимости от пола',
    color_discrete_sequence=px.colors.sequential.Viridis
)
fig.show()

График показывает, что женщины имеют более высокую вероятность выжить по сравнению с мужчинами. Это может свидетельствовать о том, что на корабле действовали некоторые правила, позволяющие женщинам и детям спасаться первыми.
Пол, очевидно, является важным признаком для классификации, и его стоит учитывать при построении моделей.

Посмотрим, как влияет цена билета на наличие `Cabin` каюты\кабины

In [22]:
# Строим график зависимости наличия кабины от цены билета
fig = px.box(
    train_df,
    x='Cabin',
    y='Fare',
    labels={'Cabin': 'Наличие кабины (0 = Нет, 1 = Да)', 'Fare': 'Цена билета'},
    title='Зависимость наличия кабины от цены билета'
)

fig.show()

как видно, в среднем каюта была у тех, у кого билет был дороже.

**Влияние наличии каюты на выживаемость(`Cabin`)**


Посмотрим, как наличие каюты влияет на выживаемость:

In [23]:
# Строим график зависимости выживаемости от наличия кабины
fig = px.bar(
    train_df.groupby('Cabin')['Survived'].mean().reset_index(),
    x='Cabin',
    y='Survived',
    labels={'Cabin': 'Наличие кабины (0 = Нет, 1 = Да)', 'Survived': 'Средняя выживаемость'},
    title='Зависимость выживаемости от наличия кабины'
)

fig.show()

Пассажиры с кабинами (значение 1): Средняя выживаемость в этой группе выше, что может свидетельствовать о том, что пассажиры с кабинами, вероятно, принадлежали к более высокому классу, где условия были более благоприятными, что повышало их шансы на выживание.

Пассажиры без кабины (значение 0): Для этой группы средняя выживаемость ниже, что может указывать на то, что они находились в менее защищенных местах или в более низких классах, что снижало их шансы на выживание.

Этот график подтверждает гипотезу о том, что наличие кабины могло сыграть важную роль в выживаемости пассажиров Титаника.

**Выживаемость в зависимости от класса (`Pclass`)**

Анализируем, как класс влияет на вероятность выживания:

In [24]:
# Группируем данные по классу и считаем среднюю выживаемость
pclass_survived = train_df.groupby('Pclass')['Survived'].mean().reset_index()

# Создаем интерактивный график
fig = px.bar(
    pclass_survived,
    x='Pclass',
    y='Survived',
    color='Pclass',
    labels={'Pclass': 'Класс', 'Survived': 'Доля выживших'},
    title='Выживаемость в зависимости от класса',
    color_discrete_sequence=px.colors.sequential.Viridis
)
fig.show()

В зависимости от класса пассажира заметно, что вероятность выживания снижается от первого класса к третьему. Это подтверждает предположение о том, что пассажиры первого класса имели лучшие условия и более высокий приоритет в спасении.
Этот признак стоит учитывать при обучении модели классификации.

**Корреляции между признаками**

Построим матрицу корреляций для анализа связей между числовыми признаками:

In [25]:
# Рассчитываем матрицу корреляций
corr_matrix = train_df.corr()

# Создаем интерактивную матрицу корреляций
fig = go.Figure(
    data=go.Heatmap(
        z=corr_matrix.values,
        x=corr_matrix.columns,
        y=corr_matrix.columns,
        colorscale='Viridis',
        colorbar=dict(title="Корреляция")
    )
)
fig.update_layout(title='Матрица корреляций', xaxis=dict(title='Признаки'), yaxis=dict(title='Признаки'))
fig.show()

Матрица корреляций помогает оценить, какие признаки имеют сильные связи с выживаемостью. Например, каюта и пол имеют высокую корреляцию с выживаемостью.
Важно обратить внимание на корреляции между признаками, чтобы выбирать только значимые переменные при обучении моделей.

### Выводы

Исходя из анализа, мы потдвердили гипотезу о том, что на выживаемость сильно влияет пол(женский $\Rightarrow$ больший шанс выжить) и наличие каюты(наличие $\Rightarrow$ больший шанс).

# $\blacksquare$ Используя любой из классификаторов получить значения `precision`, `recall` и `accuracy` выше, чем у моделей с занятия.
Если не удалось получить результат лучше, провести подробный анализ причины неудачи.

### Классификация `RandomForestClassifier`

In [26]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score
from sklearn.model_selection import train_test_split

# Разделим данные на признаки (X) и целевую переменную (y)
X = train_df.drop(columns=['Survived'])
y = train_df['Survived']

# Разделим данные на тренировочную и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Инициализация модели RandomForestClassifier
rf_model = RandomForestClassifier(random_state=42)

# Обучение модели
rf_model.fit(X_train, y_train)

# Предсказания на тестовой выборке
y_pred_rf = rf_model.predict(X_test)

# Оценка модели с помощью метрик
precision_rf = precision_score(y_test, y_pred_rf)
recall_rf = recall_score(y_test, y_pred_rf)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
f1_rf = f1_score(y_test, y_pred_rf)

# Выводим метрики
print(f"""RandomForestClassifier - \
Precision: {precision_rf:.2f}, \
Recall: {recall_rf:.2f}, \
Accuracy: {accuracy_rf:.2f}, \
F1 Score: {f1_rf:.2f}""")

RandomForestClassifier - Precision: 0.84, Recall: 0.70, Accuracy: 0.82, F1 Score: 0.76


Для нашего анализа метрика `accuracy` выше (и равна 0.82), чем на том, что было на семинаре(0.8). А значит, задача выполнена.

# $\blacksquare$ Используя любую из регрессионных задач получить значение `RootMeanSquaredError` ниже, чем у модели с занятия.
Если не удалось получить результат лучше, провести подробный анализ причины неудачи.

Для начала проверим наши данные:

In [27]:
# Вывод полученных датасетов
print("Тренировочный набор данных после обработки:")
display(train_df.head())

print("Тестовый набор данных после обработки:")
display(test_df.head())

Тренировочный набор данных после обработки:


PassengerId  Survived  Pclass  Sex   Age     Fare  Cabin  Embarked  Family
0            1         0       3    0  22.0   7.2500      0         2       1
1            2         1       1    1  38.0  71.2833      1         0       1
2            3         1       3    1  26.0   7.9250      0         2       0
3            4         1       1    1  35.0  53.1000      1         2       1
4            5         0       3    0  35.0   8.0500      0         2       0

Тестовый набор данных после обработки:


PassengerId  Pclass  Sex   Age     Fare  Cabin  Embarked  Family
0          892       3    0  34.5   7.8292      0         1       0
1          893       3    1  47.0   7.0000      0         2       1
2          894       2    0  62.0   9.6875      0         1       0
3          895       3    0  27.0   8.6625      0         2       0
4          896       3    1  22.0  12.2875      0         2       2

### Регрессия `LinearRegression`

В блоке кода ниже мы обучаем модель линейной регрессии для предсказания возраста пассажиров на основе других признаков. Сначала делим данные на признаки и целевую переменную (`Age`), затем разделяем данные на тренировочную и тестовую выборки. После этого стандартизируем признаки для улучшения работы модели и обучаем линейную регрессию. В конце, используя тестовую выборку, получаем предсказания и рассчитываем метрику `RMSE`, которая показывает точность модели — чем меньше значение, тем точнее модель.

In [28]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Разделим данные на признаки и целевую переменную
age_features = train_df.drop(columns=['Age', 'Survived'])  # исключаем 'Age' и 'Survived'
age_target = train_df['Age']

# Разделим данные на тренировочную и тестовую выборки
age_features_train, age_features_test, age_target_train, age_target_test = train_test_split(
    age_features, age_target, test_size=0.2, random_state=56
)

# Стандартизируем данные
scaler = StandardScaler()
age_features_train_scaled = scaler.fit_transform(age_features_train)
age_features_test_scaled = scaler.transform(age_features_test)

# Обучаем модель линейной регрессии
model = LinearRegression()
model.fit(age_features_train_scaled, age_target_train)

# Прогнозируем на тестовой выборке
age_pred = model.predict(age_features_test_scaled)

# Рассчитываем RMSE до удаления выбросов
rmse = np.sqrt(mean_squared_error(age_target_test, age_pred))
print(f"RMSE: {rmse}")

RMSE: 11.210613216523734


Здесь мы пытаемся найти выбросы в тестовой выборке, которые могут сильно искажать модель. Для этого рассчитываем отклонения между реальными значениями возраста и предсказанными. Если это отклонение больше чем 3 стандартных отклонения (как правило, выбросы расположены далеко от основного распределения), то такие значения считаются выбросами и исключаются.

In [29]:
# Определение выбросов на основе отклонений от предсказанных значений
threshold = 3 * (age_target_test - age_pred).std()
filtered_indices_test = np.abs(age_target_test - age_pred) < threshold

# Фильтрация выбросов
age_features_test_filtered = age_features_test_scaled[filtered_indices_test]
age_target_test_filtered = age_target_test[filtered_indices_test]

# Повторное обучение модели на отфильтрованных данных
model_filtered = LinearRegression()
model_filtered.fit(age_features_train_scaled, age_target_train)

# Прогнозирование на отфильтрованной тестовой выборке
age_pred_filtered = model_filtered.predict(age_features_test_filtered)

# Рассчитываем RMSE без выбросов
rmse_filtered = np.sqrt(mean_squared_error(age_target_test_filtered, age_pred_filtered))
print(f"RMSE без выбросов: {rmse_filtered}")

RMSE без выбросов: 10.468422296115952


Как видно, мы успешно оптимизировали модель и получили ошибку `RMSE` (10.47) ниже заявленой на семинаре(10.7) . Значит, задача выполнена

# $\blacksquare$ Сделать выводы о полученных результатах.

### Выводы

В ходе выполнения задачи были применены классификация и регрессия для анализа данных и прогнозирования. Для классификации использовалась модель `RandomForestClassifier`, которая показала результат с метрикой accuracy `0.82`, что выше значения, полученного на семинаре (`0.8`), и подтверждает успешное выполнение задачи, однако улучшения не настолько велики.

Для регрессии применялась `LinearRegression` для возраста пассажиров, и в результате оптимизации модели была достигнута ошибка `RMSE` `10.47`, что ниже заявленного значения `10.7` на семинаре, что также подтверждает успешное выполнение задачи.

Тем не менее, эти результаты не являются пределом возможных улучшений. Есть потенциал для дальнейшей оптимизации моделей, например, с помощью более сложных алгоритмов, а также улучшения предобработки данных и отбора признаков.